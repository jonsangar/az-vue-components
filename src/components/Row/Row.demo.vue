<template lang="pug">
  div
    Demo(title="Alineación Horizontal")
      .demo__subtitle left
      .content.align
        az-row(align="left")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle center
      .content.align
        az-row(align="center")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle right
      .content.align
        az-row(align="right")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle between
      .content.align
        az-row(align="between")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle around
      .content.align
        az-row(align="around")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle evenly
      .content.align
        az-row(align="evenly")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

    Demo(title="Alineación Vertical")
      .demo__subtitle top
      .content.valign
        az-row(valign="top")
          .block line 1
          .block line 1 <br> line 2 <br> line 3
          .block line 1 <br> line 2

      .demo__subtitle middle
      .content.valign
        az-row(valign="middle")
          .block line 1
          .block line 1 <br> line 2 <br> line 3
          .block line 1 <br> line 2

      .demo__subtitle bottom
      .content.valign
        az-row(valign="bottom")
          .block line 1
          .block line 1 <br> line 2 <br> line 3
          .block line 1 <br> line 2

      .demo__subtitle baseline
      .content.valign.baseline
        az-row(valign="baseline")
          .block(style="padding-top: 10px") base
          .block(style="padding-top: 20px") base
          .block(style="padding-bottom: 20px") base

      .demo__subtitle stretch
      .content.valign
        az-row(valign="stretch")
          .block line 1
          .block line 1 <br> line 2 <br> line 3
          .block line 1 <br> line 2

    Demo(title="Dirección")
      .demo__subtitle row
      .content
        az-row(dir="row")
          .block 1
          .block 2
          .block 3

      .demo__subtitle row-reverse
      .content
        az-row(dir="row-reverse")
          .block 1
          .block 2
          .block 3

      .demo__subtitle column
      .content
        az-row(dir="column")
          .block 1
          .block 2
          .block 3

      .demo__subtitle column-resverse
      .content
        az-row(dir="column-resverse")
          .block 1
          .block 2
          .block 3

    Demo(title="Justificación Horizontal")
      .demo__subtitle nowrap
      .content.wrap
        az-row(hwrap="nowrap")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle wrap
      .content.wrap
        az-row(hwrap="wrap")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle wrap-reverse
      .content.wrap
        az-row(hwrap="wrap-reverse")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

    Demo(title="Justificación vertical")
      .demo__subtitle top
      .content.vwrap
        az-row(hwrap="wrap" vwrap="top")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle middle
      .content.vwrap
        az-row(hwrap="wrap" vwrap="middle")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle bottom
      .content.vwrap
        az-row(hwrap="wrap" vwrap="bottom")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle between
      .content.vwrap
        az-row(hwrap="wrap" vwrap="between")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle around
      .content.vwrap
        az-row(hwrap="wrap" vwrap="around")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

      .demo__subtitle stretch
      .content.vwrap
        az-row(hwrap="wrap" vwrap="stretch")
          .block 1
          .block 2
          .block 3
          .block 4
          .block 5
          .block 6

    Demo(title="Separación Horizontal")
      .demo__subtitle none
      .content
        az-row(hgutter="none")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle small
      .content
        az-row(hgutter="small")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle default
      .content
        az-row
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle large
      .content
        az-row(hgutter="large")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

    Demo(title="Separación Verical")
      .demo__subtitle none
      .content
        az-row(vgutter="none" dir="column")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle small
      .content
        az-row(vgutter="small" dir="column")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle default
      .content
        az-row(dir="column")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

      .demo__subtitle large
      .content
        az-row(vgutter="large" dir="column")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;

    Demo(title="Responsive")
      .content
        az-row(
          dir="column"
          :palm="{dir: 'row', hgutter: 'small'}"
          :desk="{dir: 'column', hgutter: 'large'}")
          .block &nbsp;
          .block &nbsp;
          .block &nbsp;
</template>

<script>
export default {};
</script>

<style lang="scss" scoped>
.content {
  margin-bottom: half();
  overflow: hidden;
  background-color: color(gray6);
  border-radius: 3px;

  .block:nth-child(3n + 1) {
    width: 60px;
  }
  .block:nth-child(3n + 2) {
    width: 80px;
  }
  .block:nth-child(3n + 3) {
    width: 40px;
  }

  &.default .block {
    width: 100% !important;
  }
}

.block {
  width: 100%;
  font-size: font-size(small2);
  font-weight: font-weight(bold);
  color: white;
  text-align: center;
  text-overflow: ellipsis;
  white-space: nowrap;
  background-color: color(brand);
  border-radius: 3px;
}

.valign {
  position: relative;

  .row {
    height: 150px;
  }

  &.baseline::before {
    position: absolute;
    top: 2.2rem;
    left: 0;
    display: block;
    width: 100%;
    height: 1px;
    content: '';
    background-color: rgba(0, 0, 0, 0.2);
  }
}

.vwrap {
  .row {
    height: 250px;
  }

  .block {
    height: initial;
  }
}
</style>
