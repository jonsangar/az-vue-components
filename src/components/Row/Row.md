---
device: true
---

# Row

Utiliza todo el poder de flex-box para realizar una amplia gama de diseños de forma simple, sencilla y adaptable al tamaño de pantalla.

Este componente no aporta ninguna funcionalidad, solo simplifica el manejo del framework evitando tener que utilizar varias clases. Gracias a las propiedades, podemos adaptar el diseño de forma dinámica si fuese necesario.

**Por defecto el framework base incorpora todas las clases que utiliza este componente.**

### Instalación

```javascript
import { Row } from 'didor';

Vue.use(Row);
```

---

### Uso

#### Alineación Horizontal

Permite definir la alineación de los elementos a lo largo del eje horizontal. Ayuda a distribuir el espacio que queda libre en una línea cuando sus elementos tienen un tamaño fijo o han alcanzado su tamaño máximo.

```html
<az-row align="left">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row align="center">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row align="right">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row align="between">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row align="around">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row align="evenly">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>
```

#### Alineación Vertical

Permite definir como se van a alinear verticalmente los elementos.

```html
<az-row valign="top">
  <div>line 1</div>
  <div>line 1 <br> line 2 <br> line 3</div>
  <div>line 1 <br> line 2</div>
</az-row>

<az-row valign="middle">
  <div>line 1</div>
  <div>line 1 <br> line 2 <br> line 3</div>
  <div>line 1 <br> line 2</div>
</az-row>

<az-row valign="bottom">
  <div>line 1</div>
  <div>line 1 <br> line 2 <br> line 3</div>
  <div>line 1 <br> line 2</div>
</az-row>

<az-row valign="baseline">
  <div style="padding-top: 10px">base</div>
  <div style="padding-top: 20px">base</div>
  <div style="padding-bottom: 20px">base</div>
</az-row>

<az-row valign="stretch">
  <div>line 1</div>
  <div>line 1 <br> line 2 <br> line 3</div>
  <div>line 1 <br> line 2</div>
</az-row>
```

#### Dirección

Permite definir la dirección en la que se colocarán los elementos dentro del contenedor, vertical u horizontal. También permite invertir el orden en el que se alinearán los elementos.

```html
<az-row dir="row">
  <div>1</div>
  <div>2</div>
  <div>3</div>
</az-row>

<az-row dir="row-reverse">
  <div>1</div>
  <div>2</div>
  <div>3</div>
</az-row>

<az-row dir="column">
  <div>1</div>
  <div>2</div>
  <div>3</div>
</az-row>

<az-row dir="column-resverse">
  <div>1</div>
  <div>2</div>
  <div>3</div>
</az-row>
```

#### Justificación Horizontal

Por defecto, todos los elementos intentarán encajar en una línea modificando su tamaño. Puede cambiar el comportamiento con esta propiedad.

```html
<az-row hwrap="nowrap">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap-reverse">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>
```

#### Justificación Vertical

Permite indicar como se van a distribuir a lo largo del eje verticallos elementos cuando la altura del contenedor es mayor que el espacio que ocupan. **Nota:** esta propiedad solo tiene efecto cuando hay más de una línea de elementos.

```html
<az-row hwrap="wrap" vwrap="top">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap" vwrap="middle">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap" vwrap="bottom">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap" vwrap="between">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap" vwrap="around">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>

<az-row hwrap="wrap" vwrap="stretch">
  <div>1</div>
  <div>2</div>
  <div>3</div>
  <div>4</div>
  <div>5</div>
  <div>6</div>
</az-row>
```

#### Separación Horizontal

Permite indicar el espacio horizontal que separa a dos elementos.

```html
<az-row hgutter="none">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row hgutter="small">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row>
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row hgutter="large">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>
```

#### Separación Vertical

Permite indicar el espacio vertical que separa a dos elementos cuando hay varias líneas.

```pug
<az-row vgutter="none">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row vgutter="small">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row>
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>

<az-row vgutter="large">
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>
```

#### Responsive

Permite modificar algunos valores en base al tamaño de pantalla. Existe una propiedad por cada tamaño `palm`, `lap`, `small`, `desk`, `large`, donde le podemos pasar un objeto con las propiedades que son responsive:

- `dir`
- `hgutter`
- `vgutter`


```html
<az-row
  dir="column"
  :palm="{dir: 'row', hgutter: 'small'}"
  :desk="{hgutter: 'large'}"
>
  <div>...</div>
  <div>...</div>
  <div>...</div>
</az-row>
```

---

### API

| Attribute   | Description                                | Type      | Default   |
| ----------- | ------------------------------------------ | --------- | --------- |
| inline      | Determina el comportamiento del elemento: en línea o en bloque | `Boolean`  | `false` |
| grid        | Determina si los componentes hijos se tratarán como un grid      | `Boolean`  | `false` |
| align       | Determina como se alinearán los elementos hijos horizontalmente. Acepta:<br> `left` `center` `right` `between` `around` `evenly`           | `String` | --   |
| valign      | Determina como se alinearán los elementos hijos verticalmente. Acepta:<br>`stretch` `top` `middle` `bottom` `baseline`             | `String` | --    |
| dir         | Determina la dirección de la secuencia de los elementos hijos. Acepta:<br>`row` `row-reverse` `column` `column-resverse`             | `String` | --    |
| hwrap       | Determina como se justificarán los elementos hijos horizontalmente. Acepta:<br>`nowrap` `wrap` `wrap-reverse`             | `String` | --    |
| vwrap       | Determina como se justificarán los elementos hijos verticalmente. Acepta:<br>`stretch` `top` `middle` `bottom` `between` `around`           | `String` | --    |
| hgutter      | Determina la separación horizontal entre los elementos hijos. Acepta:<br>`stretch` `none` `small` `default` `large`         | `String` | --    |
| vgutter      | Determina la separación vertical entre los elementos hijos. Acepta:<br>`stretch` `none` `small` `default` `large`         | `String` | --    |
| palm<br>lap<br>small<br>desk<br>large  | Permiten modificar algunas propiedades en función del tamaño de la pantalla, las propiedades responsive son:<br>`dir` `hgutter` `vgutter`             | `Object` | --    |
