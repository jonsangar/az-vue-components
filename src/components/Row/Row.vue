<template lang="pug">
.row(
  :class="classObject")
  slot
</template>

<script>
import ResponsiveClass from '../mixins/responsiveClass';

export default {
  mixins: [ResponsiveClass],
  props: {
    inline: {
      type: Boolean,
      default: false,
    },
    align: {
      type: String,
      default: '',
      validator: value => ['left', 'center', 'right', 'between', 'around', 'evenly', ''].indexOf(value) !== -1,
    },
    valign: {
      type: String,
      default: '',
      validator: value => ['stretch', 'top', 'middle', 'bottom', 'baseline', ''].indexOf(value) !== -1,
    },
    dir: {
      type: String,
      default: '',
      validator: value => ['row', 'row-reverse', 'column', 'column-resverse', ''].indexOf(value) !== -1,
    },
    hwrap: {
      type: String,
      default: '',
      validator: value => ['nowrap', 'wrap', 'wrap-reverse', ''].indexOf(value) !== -1,
    },
    vwrap: {
      type: String,
      default: '',
      validator: value => ['stretch', 'top', 'middle', 'bottom', 'between', 'around', ''].indexOf(value) !== -1,
    },
    hgutter: {
      type: String,
      default: '',
      validator: value => ['none', 'small', 'default', 'large', ''].indexOf(value) !== -1,
    },
    vgutter: {
      type: String,
      default: '',
      validator: value => ['none', 'small', 'default', 'large', ''].indexOf(value) !== -1,
    },
    grid: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    classObject() {
      const baseClass = 'row';
      const responsiveProperties = ['dir', 'hgutter', 'vgutter'];
      const responsiveClass = this.getResponsiveClass(baseClass, responsiveProperties);

      const propertiesClass = [
        this.align ? `row--align-${this.align}` : '',
        this.valign ? `row--valign-${this.valign}` : '',
        this.dir ? `row--dir-${this.dir}` : '',
        this.hwrap ? `row--hwrap-${this.hwrap}` : '',
        this.vwrap ? `row--vwrap-${this.vwrap}` : '',
        this.hgutter ? `row--hgutter-${this.hgutter}` : '',
        this.vgutter ? `row--vgutter-${this.vgutter}` : '',
        {
          'row--inline': this.inline,
          'row--grid': this.grid,
        },
      ];

      return propertiesClass.concat(responsiveClass);
    },
  },
};
</script>
