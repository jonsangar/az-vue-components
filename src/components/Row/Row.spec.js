import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import Row from './Row.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('Row.vue', () => {
  const slotContent = '<span>foo</span>';

  const build = () => {
    const wrapper = shallowMount(Row, {
      localVue,
      slots: {
        default: slotContent,
      },
    })

    return {
      wrapper,
    }
  }

  it('Comprueba renderizado del componente Row con snapshot', () => {
    const { wrapper } = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Renderiza el contenido del slot', async () => {
    const { wrapper } = build()
    wrapper.setProps({
      inline: true,
      align: 'center',
      valign: 'middle',
      dir: 'row',
      hwrap: 'nowrap',
      vwrap: 'between',
      vgutter: 'none',
    })
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toContain(slotContent);
    expect(wrapper.html()).toMatchSnapshot()
  });

  it('Comprueba el validator para la alineación "align"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.align.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('left')).toBe(true);
    expect(validator('right')).toBe(true);
    expect(validator('center')).toBe(true);
    expect(validator('between')).toBe(true);
    expect(validator('around')).toBe(true);
    expect(validator('evenly')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakealign')).toBe(false);
  });

  it('Comprueba el validator para la alineación vertical "valign"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.valign.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('stretch')).toBe(true);
    expect(validator('top')).toBe(true);
    expect(validator('middle')).toBe(true);
    expect(validator('bottom')).toBe(true);
    expect(validator('baseline')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakevalign')).toBe(false);
  });

  it('Comprueba el validator para la dirección "dir"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.dir.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('row')).toBe(true);
    expect(validator('row-reverse')).toBe(true);
    expect(validator('column')).toBe(true);
    expect(validator('column-resverse')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakedir')).toBe(false);
  });

  it('Comprueba el validator para la envoltura horizontal "hwrap"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.hwrap.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('nowrap')).toBe(true);
    expect(validator('wrap')).toBe(true);
    expect(validator('wrap-reverse')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakehwrap')).toBe(false);
  });

  it('Comprueba el validator para la envoltura vertical "vwrap"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.vwrap.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('stretch')).toBe(true);
    expect(validator('top')).toBe(true);
    expect(validator('middle')).toBe(true);
    expect(validator('bottom')).toBe(true);
    expect(validator('between')).toBe(true);
    expect(validator('around')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakevwrap')).toBe(false);
  });

  it('Comprueba el validator para el canalón vertical "hgutter"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.hgutter.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('none')).toBe(true);
    expect(validator('small')).toBe(true);
    expect(validator('default')).toBe(true);
    expect(validator('large')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakehgutter')).toBe(false);
  });

  it('Comprueba el validator para el canalón vertical "vgutter"', async () => {
    // NOTA: Los validadores es mejor testearlos de forma aislada (sin montar);
    const validator = Row.props.vgutter.validator;

    // Testeamos el validador de la prop como una función normal
    expect(validator('none')).toBe(true);
    expect(validator('small')).toBe(true);
    expect(validator('default')).toBe(true);
    expect(validator('large')).toBe(true);
    expect(validator('')).toBe(true);
    expect(validator('fakevgutter')).toBe(false);
  });
});
