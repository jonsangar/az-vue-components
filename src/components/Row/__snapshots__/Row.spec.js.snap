// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row.vue Comprueba renderizado del componente Row con snapshot 1`] = `<div class="row"><span>foo</span></div>`;

exports[`Row.vue Renderiza el contenido del slot 1`] = `<div class="row row--align-center row--valign-middle row--dir-row row--hwrap-nowrap row--vwrap-between row--vgutter-none row--inline"><span>foo</span></div>`;
